LF-14786 Everclear [EverclearFacet v1.0.0, IEverclearFeeAdapter v1.0.0]

mirooon opened the pull request:
# Which Jira task belongs to this PR?
[LF-14786]

# Why did I implement it this way?

# Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] This pull request is as small as possible and only tackles one problem
- [ ] I have added tests that cover the functionality / test the bug
- [ ] For new facets: I have checked all points from this list: https://www.notion.so/lifi/New-Facet-Contract-Checklist-157f0ff14ac78095a2b8f999d655622e
- [ ] I have updated any required documentation

# Checklist for reviewer (DO NOT DEPLOY and contracts BEFORE CHECKING THIS!!!)

- [ ] I have checked that any arbitrary calls to external contracts are validated and or restricted
- [ ] I have checked that any privileged calls (i.e. storage modifications) are validated and or restricted
- [ ] I have ensured that any new contracts have had AT A MINIMUM 1 preliminary audit conducted on <date> by <company/auditor>


[LF-14786]: https://lifi.atlassian.net/browse/LF-14786?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mirooon committed:
added Everclear components

mirooon committed:
changes

mirooon committed:
changes

mirooon committed:
changes

mirooon committed:
updates

mirooon committed:
merge with main

mirooon committed:
updates

mirooon committed:
updates

mirooon committed:
updates

mirooon committed:
updates

mirooon committed:
Merge branch 'main' into lf-14786-everclear

mirooon committed:
Refactor EverclearFacet and IEverclearFeeAdapter: Updated SPDX license identifier, enhanced EverclearData struct with additional parameters, and added validation checks for receiver addresses. Introduced new unit tests for error handling on invalid output assets and receiver addresses

mirooon committed:
Enhance EverclearFacet documentation and implementation: Updated the Everclear protocol description, refined the EverclearData struct with detailed parameters, and added error handling for bridging to non-EVM chains. Introduced new unit tests for bridging functionality and validation of receiver addresses.

mirooon committed:
Implement native fee handling in EverclearFacet: Added nativeFee parameter to EverclearData struct, updated transaction execution to include native fees, and enhanced demo script for bridging functionality. Introduced unit tests to validate behavior with native assets and ensure proper error handling.

mirooon committed:
Remove MessageHashUtils library and replace its functionality in EverclearFacet tests: Deleted MessageHashUtils.sol, updated EverclearFacetTest to include a local implementation of the toEthSignedMessageHash function, and added a MockEverclearFeeAdapter for testing native fee handling.

mirooon committed:
deployed EverclearFacet on staging

mirooon committed:
updates

mirooon committed:
Update deployment logs and demo script for EverclearFacet: Changed deployed address in JSON logs, added support for Non-EVM chain handling in demo script, and refined ABI decoding logic for better compatibility with both EVM and Non-EVM chains

mirooon committed:
Update demoEverclear script: Added transaction hash for bridging USDC from Arbitrum to Solana to enhance logging and traceability.

mirooon committed:
update demo script

mirooon committed:
added tron script

mirooon committed:
added EverclearV2Facet

mirooon committed:
Refactor EverclearFacet and EverclearV2Facet: Updated error handling for invalid call data and unsupported chain IDs, changed SPDX license to LGPL-3.0-only in deployment scripts, and removed duplicate EverclearFacet entries from deployment JSON

mirooon committed:
Merge branch 'main' into lf-14786-everclear

mirooon committed:
added nativeFee in natspec for EverclearData

0xHarbs reviewed (commented):
Looks good

ezynda3 committed:
remove old v1 in favor of v2

ezynda3 committed:
fix sig generation

ezynda3 committed:
fix

ezynda3 committed:
docs

ezynda3 committed:
deploy to staging and create demo script

ezynda3 committed:
redeploy to staging update demo script

ezynda3 committed:
fix

ezynda3 committed:
fixes

ezynda3 committed:
Merge branch 'main' into lf-14786-everclear

ezynda3 committed:
solana

0xDEnYO commented on src/Facets/EverclearFacet.sol:
```diff
@@ -0,0 +1,218 @@
+        validateBridgeData(_bridgeData)
+        noNativeAsset(_bridgeData)
+    {
+        _bridgeData.minAmount = _depositAndSwap(
```
Since this is an intent bridge, do we not have the same issue here that we had in acrossV4?
If the swap has positive slippage and we have more input amount in the bridge, then we also need to update the `amountOutMin` parameter, otherwise the positive slippage ends up in the solver's pocket. 
Has this been considered?

ezynda3 replied:
✅ Fixed in commit 06ff1a2

Implemented positive slippage handling following the AcrossFacetV4 pattern:
- Added MULTIPLIER_BASE constant (1e18)
- Added amountOutMinMultiplier field to EverclearData struct
- Recalculate amountOutMin after swaps using multiplier

This ensures users benefit from favorable swap execution rather than solvers keeping the difference. The multiplier accounts for decimal differences between input and output tokens (calculated as: multiplierPercentage * 1e18 * 10^(outputDecimals - inputDecimals)).

0xDEnYO commented on test/solidity/Facets/EverclearFacet.t.sol:
```diff
@@ -0,0 +1,1102 @@
+import { InvalidCallData, InvalidConfig, InvalidNonEVMReceiver, InvalidReceiver, NativeAssetNotSupported } from "lifi/Errors/GenericErrors.sol";
+
+// Stub EverclearFacet Contract
+contract TestEverclearFacet is EverclearFacet {
```
Use TestWhitelistManagerBase

0xDEnYO commented on test/solidity/Facets/EverclearFacet.t.sol:
```diff
@@ -0,0 +1,1102 @@
+    //       })
+    //     }
+    //   )
+    uint256 internal usdCAmountToSend = 99934901; // its defaultUSDCAmount - fee (100000000 - 65099)
```
would be cleaner to just calculate it instead of hardcoding (and then having to explain it)

0xDEnYO commented on script/demoScripts/demoEverclear.ts:
```diff
@@ -0,0 +1,575 @@
+ * @param ethPrivateKey - Ethereum private key (with or without 0x prefix)
+ * @returns Solana address in base58 format
+ */
+function deriveSolanaAddress(ethPrivateKey: string): string {
```
probably makes sense to move this to demoscripthelpers (same for function below)?

ezynda3 replied:
✅ Fixed in commit a450a9d

Moved deriveSolanaAddress() and solanaAddressToBytes32() functions to script/demoScripts/utils/demoScriptHelpers.ts for better reusability across Solana-related scripts.

ezynda3 committed:
fixes

ezynda3 committed:
Add EverclearFacet entry to deployRequirements.json

- Add _feeAdapter config pointing to everclear.json
- Required for deployment script validation

Addresses review comment GH-2579265264

ezynda3 committed:
Update EverclearFacet version to 1.0.0

- Change from 2.0.0 to 1.0.0 since this is the initial release
- Version 1.0.0 was never audited/deployed

Addresses review comment GH-2579257364

ezynda3 committed:
Implement positive slippage handling for Everclear swaps

- Add MULTIPLIER_BASE constant for decimal adjustments
- Add amountOutMinMultiplier field to EverclearData struct
- Recalculate amountOutMin after swaps to capture positive slippage
- Update _startBridge to accept memory parameter instead of calldata
- Update TestEverclearFacet to inherit TestWhitelistManagerBase
- Replace deprecated addDex/setFunctionApprovalBySignature methods

Addresses review comments GH-2579279199 and GH-2579299731

ezynda3 committed:
Fix fee validation to prevent zero bridge amount

- Change condition from < to <= in fee validation
- Prevents case where fee equals minAmount resulting in 0 bridge amount

Addresses review comment GH-2579281402

ezynda3 committed:
Clarify fee handling comments in EverclearFacet

- Update comments to specify FeeAdapter pulls from diamond, not user
- Makes it clear the user has already deposited to diamond

Addresses review comment GH-2579294941

ezynda3 committed:
Improve test code quality and add event checks

- Replace hardcoded values with calculated usdCAmountToSend
- Add event emission checks to testBase_CanBridgeTokens
- Add event emission checks to test_CanBridgeTokensToNonEVMChain
- Remove skip_ prefix to enable test_CanBridgeTokensWithNativeFee
- Clarify error handling comment for FeeAdapter revert
- 31/33 tests passing (2 swap tests need signature updates)

Addresses review comments GH-2579302938, GH-2579316574, GH-2579318141,
GH-2579316882, GH-2579417235, GH-2579421315

ezynda3 committed:
Clean up demo script and add amountOutMinMultiplier

- Remove AI confirmation comments and architecture section
- Add amountOutMinMultiplier field to everclearData struct

Addresses review comment GH-2579424181

ezynda3 committed:
Move Solana utility functions to demoScriptHelpers

- Move deriveSolanaAddress and solanaAddressToBytes32 to helpers module
- Add @solana/web3.js import to demoScriptHelpers
- Export functions for reuse in other Solana-related scripts
- Update demoEverclear.ts to import from helpers

Addresses review comment GH-2579431874

ezynda3 reviewed (commented):
The 4 chains (mainnet, base, optimism, arbitrum) represent our initial deployment scope focusing on the most popular networks. While Everclear supports 22 chains total, this is an intentional business decision for the initial release. We can expand to additional Everclear-supported chains in future deployments as needed.

ezynda3 commented on src/Facets/EverclearFacet.sol:
```diff
@@ -0,0 +1,218 @@
+/// @title EverclearFacet
+/// @author LI.FI (https://li.fi)
+/// @notice Provides functionality for bridging through Everclear
+/// @custom:version 2.0.0
```
✅ Fixed in commit e12a85e

Updated @custom:version from 2.0.0 to 1.0.0 since this is the initial release (version 1.0.0 was never audited/deployed).

ezynda3 commented on src/Facets/EverclearFacet.sol:
```diff
@@ -0,0 +1,218 @@
+                _everclearData.receiverAddress,
+                _bridgeData.sendingAssetId,
+                _everclearData.outputAsset,
+                _bridgeData.minAmount - _everclearData.fee, // fee is deducted from the minAmount and it's pulled from the sender separately
```
✅ Fixed in commit 55a7b9d

Updated comments to clarify that "sender" refers to the diamond contract, not the user. Changed to: "bridge amount (fee is deducted and pulled by FeeAdapter from the diamond separately)"

ezynda3 commented on test/solidity/Facets/EverclearFacet.t.sol:
```diff
@@ -0,0 +1,1102 @@
+        assertBalanceChange(
+            ADDRESS_USDC,
+            USER_SENDER,
+            -int256(99934901 + validEverclearData.fee) // 99934901 is defaultUSDCAmount - fee (100000000 - 65099)
```
✅ Fixed in commit ea23948

Replaced hardcoded values with usdCAmountToSend variable at lines 310 and 334.

ezynda3 committed:
Revert positive slippage handling changes

- Remove amountOutMinMultiplier field from EverclearData struct
- Remove MULTIPLIER_BASE constant
- Restore original swap+bridge flow without amountOutMin recalculation
- Revert _startBridge parameter from memory to calldata

The positive slippage handling approach is incompatible with Everclear's
signature-based validation. The API generates signatures that include
amountOutMin, which cannot be modified after the signature is created.
This needs further discussion with the Everclear team.

All tests now passing: 33/33

Reverts part of commit 06ff1a2

ezynda3 committed:
Implement positive slippage protection for Everclear swaps

- Capture positive slippage from pre-bridge swaps
- Transfer excess amount back to user instead of adjusting amountOutMin
- This approach maintains signature validity (unlike AcrossV4 multiplier)
- Follows the same pattern as EcoFacet

The key difference from the reverted approach:
- OLD: Tried to adjust amountOutMin (broke signature validation)
- NEW: Send excess directly to user, bridge expected amount only

This ensures users receive the benefit of favorable swaps while
maintaining compatibility with Everclear's signature-based validation.

All tests passing: 33/33

Addresses review comment GH-2579279199

ezynda3 committed:
Merge branch 'main' into lf-14786-everclear

ezynda3 committed:
Improve error comment to specify EVM OutOfFunds error

- Clarify that the revert is an EVM-level OutOfFunds error
- Occurs when trying to forward more native value than was sent
- Generic vm.expectRevert() is appropriate for low-level EVM errors

Addresses review comment GH-2579421315

ezynda3 committed:
Add all supported Everclear chains with FeeAdapter addresses

- Added 18 additional chains beyond initial 4
- All chains verified to exist in networks.json
- FeeAdapter addresses sourced from Everclear docs
- Chains: bsc, unichain, polygon, zksync, ronin, apechain, mode,
  avalanche, linea, blast, scroll, taiko, berachain, gnosis,
  mantle, sonic, ink

Addresses review comment GH-2579252980

ezynda3 committed:
Fix test signature format: use V2 FeeAdapter signature helpers

Replace manual 4-parameter signature creation with proper createEverclearV2Signature helper in:
- testRevert_InsufficientNativeFee (GH-2581029572)
- testRevert_SwapAndBridgeInsufficientNativeFee (GH-2581029580)

These tests were using simplified signature formats that would fail on signature validation instead of testing the intended insufficient native fee scenarios. Now properly test the native fee validation logic.

ezynda3 committed:
Update IEverclearFeeAdapter version to 1.0.0

Consistent with EverclearFacet - this is the initial release, not 2.0.0

0xDEnYO commented on src/Facets/EverclearFacet.sol:
```diff
@@ -0,0 +1,235 @@
+            actualAmountAfterSwap = _bridgeData.minAmount;
+        }
+
+        _bridgeData.minAmount = actualAmountAfterSwap;
```
actualAmountAfterSwap cannot be lower than _bridgeData.minAmount since _depositAndSwap function will ensure the minAmount.

So we have two cases: 
Case 1: `actualAmountAfterSwap == _bridgeData.minAmount` (no slippage)
Case 2: `actualAmountAfterSwap > _bridgeData.minAmount` (positive slippage)

in both cases `_bridgeData.minAmount` should remain the same, I dont think it makes sense to update it. 
Also line 136 can be removed imho.

0xDEnYO commented on src/Facets/EverclearFacet.sol:
```diff
@@ -0,0 +1,235 @@
+        // If we got more than expected from the swap (positive slippage),
+        // send the extra back to the user. We cannot adjust amountOutMin
+        // because Everclear's signature validation includes the original value.
+        if (actualAmountAfterSwap > _bridgeData.minAmount) {
```
There is question here that I need to check with BE team. Please lets keep this open until I resolve it.

mirooon committed:
updated demoEverclear.ts

ezynda3 committed:
Remove redundant variable assignments in positive slippage handling

ezynda3 committed:
Merge branch 'main' into lf-14786-everclear

ezynda3 committed:
add refundReceiver
